implementations/rust/ockam/ockam_vault/src/storage/persistent_storage.rs

```diff
@@ -0,0 +1,288 @@
+use crate::{constants, Secret, SecretAttributes, SecretType, StoredSecret};
+use ockam_core::compat::boxed::Box;
+use ockam_core::compat::sync::Arc;
+use ockam_core::{async_trait, KeyId, Result};
+use ockam_node::{FileValueStorage, InMemoryKeyValueStorage, KeyValueStorage, ValueStorage};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::BTreeMap;
+use std::path::Path;
+
+/// Storage for a Vault data backed by a file
+/// The `FileValueStorage` implementation takes care of locking / unlocking the underlying file
+/// in the presence of concurrent accesses
+pub struct PersistentStorage {
+    storage: Arc<FileValueStorage<StoredSecrets>>,
+    cache: Arc<dyn KeyValueStorage<KeyId, StoredSecret>>,
```

Comment:
can't we implement a neutral cache taking any `KeyValueStorage` as an argument?

Reply:
I don't see what that would look like with well-typed values? Could you sketch what you have in mind?

Reply:
```
struct StorageCache<K,V>{
    cache: Arc<dyn KeyValueStorage<K, V>>,
    storage: Arc<dyn KeyValueStorage<K, V>>,
}

impl StorageCache {
  fn create(store_to_cache: Arc<dyn KeyValueStorage<K, V>>) -> Self {
   Self{
     cache: Default::default(),
     store: store_to_cache,
     }
   }
}

impl KeyValueStorage for StorageCache {
   async fn get(&self, key: &K) -> Result<Option<V>> {
     if let Some(x) = self.cache.get(key) {
       return x;
      }
      if let Some(x) = self.proxied.get(key) {
        self.cache.put(key,x);
        return x;
      }
      return None;
  }
  ...
}
```

Reply:
I _think_ I see what you mean. Are you hinting at the fact that `PersistentStorage::storage` could already have its own cache, as part of its implementation? The problem is that this is a `FileValueStorage` not a `FileKeyValueStorage` for now, so it can not integrate a key / value cache since it has no notion of a key.